src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+        }
+
+        internal Stream UnderlyingStream
+        {
+            [Pure]
+            get
+            { return _stream; }
+        }
+
+        internal Int32 BufferSize
+        {
+            [Pure]
+            get
+            { return _bufferSize; }
+        }
+
+        public override bool CanRead
+        {
+            [Pure]
+            get
+            { return _stream != null && _stream.CanRead; }
+        }
+
+        public override bool CanWrite
+        {
+            [Pure]
+            get
+            { return _stream != null && _stream.CanWrite; }
+        }
+
+        public override bool CanSeek
+        {
+            [Pure]
+            get
+            { return _stream != null && _stream.CanSeek; }
+        }
+
+        public override Int64 Length
+        {
+            get
+            {
+                EnsureNotClosed();
+
+                if (_writePos > 0)
+                    FlushWrite();
+
+                return _stream.Length;
+            }
+        }
+
+        public override Int64 Position
+        {
+            get
+            {
+                EnsureNotClosed();
+                EnsureCanSeek();
+
+                Contract.Assert(!(_writePos > 0 && _readPos != _readLen), "Read and Write buffers cannot both have data in them at the same time.");
+                return _stream.Position + (_readPos - _readLen + _writePos);
+            }
+            set
+            {
+                if (value < 0)
+                    throw new ArgumentOutOfRangeException("value", SR.ArgumentOutOfRange_NeedNonNegNum);
+                Contract.EndContractBlock();
+
+                EnsureNotClosed();
+                EnsureCanSeek();
+
+                if (_writePos > 0)
+                    FlushWrite();
+
+                _readPos = 0;
+                _readLen = 0;
+                _stream.Seek(value, SeekOrigin.Begin);
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing && _stream != null)
+                {
+                    try
+                    {
+                        Flush();
+                    }
+                    finally
+                    {
+                        _stream.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                _stream = null;
+                _buffer = null;
+
+                // Call base.Dispose(bool) to cleanup async IO resources
+                base.Dispose(disposing);
+            }
+        }
+
+        public override void Flush()
+        {
+            EnsureNotClosed();
+
+            // Has WRITE data in the buffer:
+            if (_writePos > 0)
+            {
+                FlushWrite();
+                Contract.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
+                return;
+            }
+
+            // Has READ data in the buffer:
```

Comment:
WRITE is also capitalized... I guess to get the point across? I'll normalize them.